### PR TITLE
Added draft for a console interface

### PR DESCRIPTION
**Pull request by danielinux**

### CODE REVIEW

File: ge.h [hunk 1]
```diff
@@ -189,4 +292,6 @@ struct pulse_event {
 /* Defined in pulse.c: execute pulse events */
 int pulse(struct ge *ge);
 
+int ge_struct_sizeof(void);
```
Comment (biappi) [inline ge.h:295]: this is unneeded with the dgram communication, it can be nuked from orbit

File: ge.c [hunk 2]
```diff
@@ -129,3 +150,8 @@ int ge_run(struct ge *ge)
 
     return 0;
 }
+
+int ge_struct_sizeof(void)
```
Comment (biappi) [inline ge.c:154]: same as below

File: msl-states.c [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    { TI05, CI08, 0 },
+    { TI06, CI76, 0 },
+    { TI06, CI80, 0 },
+    { TI06, CI81, 0 },
+    { TI06, CI82, 0 },
+    { TI06, CU01, state_80_TI06_CU01 },
+    { TI06, CU03, state_80_TI06_CU03 },
+    { TI06, CU05, state_80_TI06_CU05 },
+    { TI06, CU06, 0 },
+    { MAX_CLOCK, 0 }
```
Comment (biappi) [inline msl-states.c:31]: shouldn't it be `{ MAX_CLOCK, 0, 0 }`?

however if we're OK with implicit inits, we can remove the trailing `, 0` in states too!

**Comment by biappi** [comment]
regarding the refactoring: i see you moved the implementations to their own c files, is there a specific reason for it? in this way we're required to needlessly duplicate the prototypes

File: msl.c [hunk 4]
```diff
@@ -7,23 +7,22 @@ struct msl_timing_state* msl_get_state(uint8_t SO)
...
+    do {
+        chart = &state->chart[i++];
         if (chart->clock == ge->current_clock) {
             if (chart->condition && !chart->condition(ge))
                 continue;
-
             chart->command(ge);
         }
-    }
+    } while (chart->clock < MAX_CLOCK);
```
Comment (rizlik130) [inline msl.c:27]: I've no preference between an explicit number of elements and a sentinel. 
I would rename MAX_CLOCK in END_OF_STATUS.

File: console_socket.c [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    struct sockaddr_un dst;
+    int ret;
+    socklen_t ssz = sizeof(struct sockaddr_un);
+    if (ge->ge_console_socket < 0) {
+        return -1;
+    }
+    ge->console.lamps.RO = ge->rRO;
+    ge->console.lamps.SO = ge->rSO;
+    ge->console.lamps.SA = ge->rSA;
+    ge->console.lamps.FA = ge->rFA & 0x0F;
```
Comment (rizlik130) [inline console_socket.c:39]: I think updating the lamps of the console should be done in ge.c run_cycle as it's general. In the future a console not using sockets may want the updated lamps.

File: ge.c [hunk 6]
```diff
@@ -13,6 +15,12 @@ static int ge_halted(struct ge *ge)
 
 int ge_init(struct ge *ge)
 {
+    ge->ge_console_socket = console_socket_init();
```
Comment (rizlik130) [inline ge.c:18]: I would prefer a middle layer, with peris that register init/on_cycle and deinit functions callbacks

File: ge.c [hunk 7]
```diff
@@ -101,11 +114,19 @@ int ge_run_cycle(struct ge *ge)
...
     for(ge->current_clock = TO00; ge->current_clock < MAX_CLOCK; ge->current_clock++) {
+
         /* Execute machine logic for pulse*/
         pulse(ge);
 
         /* Execute the commands from the timing charts */
         msl_run_state(ge, state);
+
+        /* Update console socket */
+        console_socket_check(ge);
```
Comment (rizlik130) [inline ge.c:125]: as above

File: ge.h [hunk 8]
```diff
@@ -152,12 +239,28 @@ struct ge {
 
     uint8_t AINI:1;
     /* Faults: TODO (pp. 139-141) */
+    uint8_t rFA;
 
     struct ge_console console;
+    int ge_console_socket;
```
Comment (rizlik130) [inline ge.h:245]: I don't think this should be here.
This should be internal to the console_socket implementation

File: ge.h [hunk 9]
```diff
@@ -152,12 +239,28 @@ struct ge {
...
 
     /* Memory */
     uint8_t mem[MEM_SIZE];
+
+
+    int step_by_step:1; /* XXX: replace with signal name */
+    int operator_call:1; /* XXX: replace with signal name */
+
+    int JS1:1; /* console jump condition */
+    int JS2:1; /* console jump condition */
```
Comment (rizlik130) [inline ge.h:256]: are this FF conntected to some signal?

File: ge.h [hunk 10]
```diff
@@ -27,14 +27,101 @@ enum clock {
...
+            B_SWITCH_1:1,
+            B_SWITCH_2:1,
+            B_STEP_BY_STEP:1,
+            B_LOAD_1_2:1,
+            B_EMERGEN_OFF:1,
+            B_STANDBY:1,
+            B_MEM_CHECK:1,
+            B_CLEAR:1,
+            B_LOAD:1,
+            B_HALT_START:1,
```
Comment (rizlik130) [inline ge.h:118]: nit: what's the main reason for the UPCASE?

**Review by rizlik130 — APPROVED** [review]

**Pull request merged**